modules/contact/include/constraints/ComputeFrictionalForceLMMechanicalContact.h

```diff
@@ -57,12 +65,29 @@ class ComputeFrictionalForceLMMechanicalContact : public ComputeWeightedGapLMMec
   /// A map from node to weighted gap
   std::unordered_map<const DofObject *, ADReal> _dof_to_weighted_tangential_velocity;
 
+  /// A map from node to weighted gap for an additional tangential direction (three dimensions)
+  std::unordered_map<const DofObject *, ADReal> _dof_to_weighted_tangential_velocity_dir;
+
   /// A pointer member that can be used to help avoid copying ADReals
   const ADReal * _tangential_vel_ptr = nullptr;
 
+  /// A pointer member that can be used to help avoid copying ADReals (for 3D)
+  const ADReal * _tangential_vel_dir_ptr = nullptr;
+
   /// The value of the tangential velocity at the current quadrature point
   ADReal _qp_tangential_velocity;
 
+  /// The value of the tangential velocity at the current quadrature point
+  /// in an additional direction, required for three-dimensional cases
+  ADReal _qp_tangential_velocity_dir;
+
+  /// The value of the tangential velocity at the current node
+  ADRealVectorValue _qp_tangential_velocity_nodal;
+
+  /// The value of the tangential velocity at the current node
+  /// in an additional direction, required for three-dimensional cases
+  ADRealVectorValue _qp_tangential_velocity_dir_nodal;
```

Comment:
Begs the same question here. A `_dir` suffix doesn't really communicate to me what the data member is for, e.g. that it's for holding an additional direction when in 3D vs. 2D. I would vote to make the original data members `std::vectors` and size them in the constructor according to the mesh dimension